### PR TITLE
✨  Micro App for Assessment Delivery 

### PR DESCRIPTION
**Pull request by Beulah-Matt**
# Description
This PR  allows for more comprehensive assessments by: 
 - Having a micro-app block on the bot editor to add an assessment link. 
 - Write a handler for WhatsApp call to action button API that will be clicked on by a user to be redirected to a micro-app
 - Register Goomza Micro-app on Firebase hosting and edit firebase.json config
 - Have a micro-app assessment page which renders questions as per content creator's design preferences. 
 - The page has a space for a logo and is Goomza branded. 
 - Navigation from section to section is dependent on meeting route params specifications. 
 - Handler to update a learner's progress in the case of an assessment as they move from page to page.
 
Fixes # [CLM-440]

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

# Screenshot (optional)

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

[CLM-440]: https://elewake.atlassian.net/browse/CLM-440?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/conv-lm-backend/src/main.ts [hunk 1]
```diff
@@ -41,3 +41,7 @@ export * from './app/ai-agent/intent/create-intent.function';
 export * from './app/ai-agent/intent/delete-intent.function';
 export * from './app/ai-agent/intent/edit-intent.function';
 export * from './app/ai-agent/intent/get-intent.function';
+
+
+//  Micro Apps Base
+export * from './app/micro-apps/base/init-microapp.function'
```
Comment (jrosseel) [inline apps/conv-lm-backend/src/main.ts:47]: You created two fns but only one is exported. Progress one is not exported?

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 2]
```diff
@@ -0,0 +1,891 @@
+{
+  "ELS": {
+    "TOGGLES": {
+      "SWITCH-LANG": "Changer de langue"
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:4]: Langue

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+    }
+  },
+  "LOGIN": {
+    "LOGIN": {
+      "LOGIN": "Connexion",
+      "OR": "ou",
+      "CREATE-ACCOUNT": "Créer un compte",
+      "LOGIN-FORM": {
+        "FLD-EMAIL": "Email",
+        "FLD-EMAIL-ERR": "Veuillez fournir une adresse électronique valide",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:14]: L'email n'est pas dans un format correcte

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+        "FLD-PASSWD": "Mot de passe",
+        "FLD-REMEMBER-ME": "Se souvenir de moi",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:16]: Souviner

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+        "LOCATION": "Location"
+      }
+    },
+    "BLOCK": {
+      "MENU": {
+        "EDIT": "Modifier",
+        "COPY": "Copie",
+        "DELETE": "Supprimer"
+      },
+      "BLOCK_ID": "ID de bloc",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:72]: Bloc ID

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+      "TITLES": {
+        "IMAGE": "Image Bloquer",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:74]: Remove Bloquer everywhere

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+        "PHONE": "Téléphoner Bloquer",
+        "MESSAGE": "Message Bloquer",
+        "QUESTION": "Question: Boutons",
+        "AUDIO": "l'audio Bloquer",
+        "VIDEO": "Video Bloquer",
+        "STICKER": "Autocollant Bloquer",
+        "LIST": "Liste Bloquer",
+        "DOCUMENT": "Document Bloquer",
+        "WEBHOOK": "webhook Bloquer",
+        "IMAGE-INPUT": "Entrée d'images",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:87]: Télécharger d'Images

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 8]
```diff
@@ -0,0 +1,891 @@
...
+        "VIDEO-INPUT": "entrée vidéo",
+        "AUDIO-INPUT": "entrée audio",
+        "LOCATION-INPUT": "saisie de position",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:90]: Télécharger ma position (Pin)

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json (hunk 7 above)
Comment (jrosseel) [reply apps/conv-lm-micro-app/src/assets/i18n/fr.json:87]: entree = Télécharger

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 9]
```diff
@@ -0,0 +1,891 @@
...
+    "BLOCK": {
+      "MENU": {
+        "EDIT": "Modifier",
+        "COPY": "Copie",
+        "DELETE": "Supprimer"
+      },
+      "BLOCK_ID": "ID de bloc",
+      "TITLES": {
+        "IMAGE": "Image Bloquer",
+        "LOCATION": "Emplacemant Bloquer",
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:75]: Pin du position

File: apps/conv-lm-micro-app/src/assets/i18n/fr.json [hunk 10]
```diff
@@ -0,0 +1,891 @@
+{
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/i18n/fr.json:1]: Why does the micro-app have all translations of the full app? Please trim down to what it needs!

File: apps/conv-lm-micro-app/src/assets/icons/settings-stroked.svg
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/assets/icons/settings-stroked.svg:1]: Only keep relevant images to this app

File: apps/conv-lm-micro-app/src/styles.scss [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+.wide {
+  padding: 1rem;
+  width: 15rem;
+}
+
+.frame-connector {
+  z-index: -1;
+}
+
+.jtk-overlay {
```
Comment (jrosseel) [inline apps/conv-lm-micro-app/src/styles.scss:50]: Is this relevant to the micro-app? Please remove if not

File: libs/features/convs-mgr/conversations/assessments/src/lib/components/assessment-card/assessment-card.component.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  @Input() assessmentForm: FormGroup;
+
+  /** Method called to track progress */
+  @Input() progressCallback: () => void;
+
+  stepperForm = true
+
+  /** Tracking questions using stepper */
+  currentStep = 0;
+  totalSteps  = 0;
```
Comment (jrosseel) [inline libs/features/convs-mgr/conversations/assessments/src/lib/components/assessment-card/assessment-card.component.ts:24]: EACH attribute nbeeds a JS Doc

File: libs/features/convs-mgr/conversations/assessments/src/lib/utils/calculate-progress.util.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import { FormArray } from "@angular/forms";
+
+/**
+ * A function that calculates how far a learner is in answering questions. 
+ * @param controls form array controls
+ * @returns a percentage tracker
+ */
+export function __CalculateProgress(formArray: FormArray): number
```
Comment (jrosseel) [inline libs/features/convs-mgr/conversations/assessments/src/lib/utils/calculate-progress.util.ts:8]: Good job in terms of both comenting, function name and keeping a lean fn

File: libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  
+    this._microApp$$.next({
+      appId: this.app.appId,
+      endUserId: this.app.endUserId,
+      config: this.app.config,
+      startedOn: Date.now(),
+      finishedOn: '',
+    } as unknown as MicroAppStatusTypes);
+  
+    this._microAppStatusServ.setMicroAppSections(mainSection);
```
Comment (jrosseel) [inline libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.ts:68]: A clarifying comment on why you're calling each of these setMicro app things will help

File: libs/features/micro-app/src/lib/pages/platform-redirect-page/platform-redirect-page.component.html [hunk 15]
```diff
@@ -0,0 +1,10 @@
+<div class="page-container">
+  <div class="image">
+    <img src="/assets/icons/right_direction.svg" />
+  </div>
+  <div class="feedback">
+    <div class="feedback-text">
+       <p> Redirecting you to back to the platform in {{COUNTDOWN}} seconds</p>
```
Comment (jrosseel) [inline libs/features/micro-app/src/lib/pages/platform-redirect-page/platform-redirect-page.component.html:7]: Multi-lang

File: libs/functions/bot-engine/main/src/lib/io/outgoing-message.parser.ts [hunk 16]
```diff
@@ -35,6 +35,8 @@ export abstract class OutgoingMessageParser
 
   // abstract getLocationBlockParserOut (storyBlock: StoryBlock, phone: string): Message
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/io/outgoing-message.parser.ts:36]: Needs a JSdoc, and perhapsa better name :)

File: libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+import { MicroApp, MicroAppStatus, MicroAppStatusTypes } from "@app/model/convs-mgr/micro-app/base";
+
+import { IProcessOperationBlock } from "../models/process-operation-block.interface";
+
+import { BlockDataService } from "../../data-services/blocks.service";
+import { ConnectionsDataService } from "../../data-services/connections.service";
+import { EndUserDataService } from "../../data-services/end-user.service";
+import { ActiveChannel } from '../../../model/active-channel.service';
+
+export class MicroAppBlockService implements IProcessOperationBlock
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts:18]: Comment

File: libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+{
+  sideOperations: Promise<any>[] = [];
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts:20]: Comments

File: libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    private blockDataService: BlockDataService, 
+    private connDataService: ConnectionsDataService, 
+    private tools: HandlerTools,
+    private _activeChannel: ActiveChannel
+  )
+  { }
+
+  public async handleBlock(storyBlock: MicroAppBlock, updatedCursor: Cursor, orgId: string, endUser: EndUser)
+  {
+    // 1. Create the interactive btn block - return this + updated cursor
```
Comment (jrosseel) [inline libs/functions/bot-engine/main/src/lib/services/process-operation-block/block-type/micro-app-block.service.ts:34]: Move this 1. -> 4.1 to JsDock of handleBlock

File: libs/functions/bot-engine/messenger/src/lib/io/outgoing-message-parser/messenger-api-outgoing-message-parser.class.ts [hunk 20]
```diff
@@ -131,6 +131,23 @@ export class MessengerOutgoingMessageParser extends OutgoingMessageParser
   {
     throw new Error('Method not implemented.');
   }
+
+  getInteractiveUrlButtonParserOut(buttonBlockBlock: InteractiveURLButtonBlock , recepientId: string): any
```
Comment (jrosseel) [inline libs/functions/bot-engine/messenger/src/lib/io/outgoing-message-parser/messenger-api-outgoing-message-parser.class.ts:135]: Add comment. It/s not clear what interactiveUrlButton is

File: libs/model/convs-mgr/conversations/admin/system/src/lib/event.interface.ts [hunk 21]
```diff
@@ -58,13 +60,15 @@ export interface Cursor extends IObject
    */
   parentStack?: RoutedCursor[];
 
+  microappStack?: MicroAppStatus[];
```
Comment (jrosseel) [inline libs/model/convs-mgr/conversations/admin/system/src/lib/event.interface.ts:63]: Remove

File: libs/model/convs-mgr/functions/src/lib/whatsapp/whatsapp-interactive-message.interface.ts [hunk 22]
```diff
@@ -15,6 +15,20 @@ interface InteractiveInfo {
   body:{ text: string },
 }
 
+export interface InteractiveURLMessage extends InteractiveInfo {
```
Comment (jrosseel) [inline libs/model/convs-mgr/functions/src/lib/whatsapp/whatsapp-interactive-message.interface.ts:18]: add comments

File: libs/model/convs-mgr/micro-app/base/src/lib/micro-app-progress.interface.ts [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+  appId: string;
+  /** Id of the user engaging with the app */
+  endUserId: string;
+  /** ID of an organization */
+  orgId: string;
+  /** Current question a user is in */
+  milestones?: ProgressMilestones
+}
+
+export interface ProgressMilestones
```
Comment (jrosseel) [inline libs/model/convs-mgr/micro-app/base/src/lib/micro-app-progress.interface.ts:19]: Needs JsDoc

File: libs/model/convs-mgr/micro-app/base/src/lib/microapp-section-types.enum.ts [hunk 24]
```diff
@@ -0,0 +1,10 @@
+/**
+ * In a microapp, different content is rendered depending on where a user is in the microapp
+ * This interface tracks where in the microapp a user is, and renders the specific content for that position
+ */
+export enum MicroAppSectionTypes
+{
+  Start = 1,
```
Comment (jrosseel) [inline libs/model/convs-mgr/micro-app/base/src/lib/microapp-section-types.enum.ts:7]: Comment for each section

File: libs/model/convs-mgr/stories/blocks/messaging/src/lib/assessment-micro-app.interface.ts [hunk 25]
```diff
@@ -0,0 +1,15 @@
+import { MicroAppStatus } from "@app/model/convs-mgr/micro-app/base";
+import { StoryBlock } from "@app/model/convs-mgr/stories/blocks/main";
+
+
+/**
+ * The interface for an assessment specific Micro App. 
+ */
+
+export interface AssessmentMicroAppBlock extends StoryBlock {
+  name?: string;
```
Comment (jrosseel) [inline libs/model/convs-mgr/stories/blocks/messaging/src/lib/assessment-micro-app.interface.ts:10]: JsDoc

File: libs/model/convs-mgr/stories/blocks/messaging/src/lib/micro-app-block.interface.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+  /** The type of the micro-app e.g. assessment */
+  appType: MicroAppTypes;
+
+  /** Additional data for the microApp to run */
+  configs: MicroAppConfig;
+  /** Comprehensive micro-app data  */
+  appStatus?: MicroAppStatus;
+}
+
+export interface MicroAppConfig
```
Comment (jrosseel) [inline libs/model/convs-mgr/stories/blocks/messaging/src/lib/micro-app-block.interface.ts:21]: JsDoc

File: libs/private/functions/micro-apps/base/src/lib/redirect-handler.ts [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+      const result = await cursorRepo$.getDocuments(new Query().orderBy('createdOn', 'desc').limit(1));
+
+      if (!result || result.length == 0) {
+        this.responseMessage = 'User cursor empty';
+        throw this.responseMessage;
+      }
+
+      const latestCursor = result[0];
+
+      if (!latestCursor.microappStack || latestCursor.microappStack.length === 0) {
```
Comment (jrosseel) [inline libs/private/functions/micro-apps/base/src/lib/redirect-handler.ts:65]: What is this?

File: libs/private/functions/micro-apps/base/src/lib/redirect-handler.ts [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+import { Query } from '@ngfi/firestore-qbuilder';
+import { FunctionHandler, FunctionContext, RestResult } from '@ngfi/functions';
+
+import { CommunicationChannel, Cursor } from '@app/model/convs-mgr/conversations/admin/system';
+import { InitMicroAppCmd, MicroAppSectionTypes, MicroAppStatus, MicroAppStatusTypes } from '@app/model/convs-mgr/micro-app/base';
+import { ChatStatus, EndUser } from '@app/model/convs-mgr/conversations/chats';
+import { ActiveChannelFactory } from '@app/functions/bot-engine/utils';
+import { EngineBotManager } from '@app/functions/bot-engine';
+
+/**
```
Comment (jrosseel) [inline libs/private/functions/micro-apps/base/src/lib/redirect-handler.ts:11]: To review togethre!

File: libs/state/convs-mgr/micro-app/src/lib/services/micro-app-management.service.ts [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+  {
+    const initUrl = `${this._env.microAppUrl}/${INIT_MICROAPP_ENDPOINT}`
+    const payload: InitMicroAppCmd = {
+      appId
+    }
+
+    return this._http$.post<InitMicroAppResponse>(initUrl, payload);
+  }
+
+  progressCallBack(appId: string, userId: string, config: MicroApp, milestones?: ProgressMilestones)
```
Comment (jrosseel) [inline libs/state/convs-mgr/micro-app/src/lib/services/micro-app-management.service.ts:39]: JsDoc

File: libs/features/convs-mgr/conversations/assessments/src/lib/components/assessment-card/assessment-card.component.html [hunk 30]
```diff
@@ -0,0 +1,50 @@
+<div *ngIf="assessmentFormArray" >
+  <div *ngIf="!stepperForm">
+    <form [formGroup]="assessmentForm">
+      <div formArrayName="assessmentFormArray">
+        <mat-card class="question-card" *ngFor="let question of assessmentFormArray.controls; let i = index" [formGroupName]="i">
+          <mat-card-header class="question-header">
+            Question {{i + 1}}
```
[the remark this answers is not in the document]
Comment (Beulah-Matt) [reply libs/features/convs-mgr/conversations/assessments/src/lib/components/assessment-card/assessment-card.component.html:7]: Done

File: libs/features/convs-mgr/conversations/assessments/src/lib/model/assessment-question.interface.ts [hunk 31]
```diff
@@ -0,0 +1,7 @@
+export interface AssessmentQuestion{
+  id: string
```
[the remark this answers is not in the document]
Comment (Beulah-Matt) [reply libs/features/convs-mgr/conversations/assessments/src/lib/model/assessment-question.interface.ts:2]: Deleting this model since it is no longer in use.

File: libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.html [hunk 32]
```diff
@@ -0,0 +1,40 @@
...
+      <div class="text-section">
+        <div class="p-text">
+          <div class="intro">
+            <p class="main-text">{{"MICROAPPS-SCREENS.CONTENT-SECTION.INTRODUCTION-TAG" | transloco}} 
+              <span class="teal-text"> Financial Literacy!</span>
+            </p>
+          </div>
+          <div class="intro">
+            <p class="main-text">{{"MICROAPPS-SCREENS.CONTENT-SECTION.DESCRIPTION-TEXT" | transloco}} 
+              <span class="teal-text"> Personal Finance, budgeting and investment basics.</span>
```
[the remark this answers is not in the document]
Comment (Beulah-Matt) [reply libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.html:27]: This has been solved in publish assessment for Micro-Apps. 
Now taking in an assessment's description

File: libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.ts (hunk 14 above)
Comment (Beulah-Matt) [reply libs/features/micro-app/src/lib/pages/micro-app-start-page/micro-app-start-page.component.ts:68]: Removed them, setting navigation on app stage changes

File: libs/features/convs-mgr/conversations/assessments/src/lib/convs-mgr-assessments.module.ts [hunk 33]
```diff
@@ -82,8 +86,18 @@ import { AssessmentFormService } from './services/assessment-form.service';
...
+    AssessmentCardComponent,
+    AssessmentFailedSectionComponent,
+    AssessmentHeaderComponent,
+    AssessmentResultBannerComponent,
+    ContentSectionComponent,
+
     CreateAssessmentPageComponent
   ],
+  exports: [
+    ContentSectionComponent,
```
[the remark this answers is not in the document]
Comment (Beulah-Matt) [reply libs/features/convs-mgr/conversations/assessments/src/lib/convs-mgr-assessments.module.ts:103]: Resolved this by using content projection

**Comment by ReaganCn** [comment]
Update the description to cover all the changes done in this PR
